Pulumi: Migrate pyproject to PyPA standard and hatch

JimMadge opened the pull request:
## :white_check_mark: Checklist

<!--
Thank you for your pull request!

Before going any further, please review the [guidelines for contributing](../CONTRIBUTING.md) to this repository.
Replace the empty checkboxes [ ] below with checked ones [x] accordingly.
-->

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).
- [x] You have marked this pull request as a **draft** and added `'[WIP]'` to the title if needed (if you're not yet ready to merge).
- [x] You have formatted your code using appropriate automated tools (for example `./tests/AutoFormat_Powershell.ps1 -TargetPath <path to file or directory>` for Powershell).

### :arrow_heading_up: Summary

<!--
Please explain what your pull request does here.
You might (optionally) want to include screenshots here.
-->

Migrate to hatch/hatchling.

- Use hatchling for build backend
- Use hatch for development
- Use standards PEP508, PEP440, PEP621
- Demonstrate using hatch environments for, linting, testing, building, _etc._


### :closed_umbrella: Related issues

<!--
If your pull request will close any open issues (hopefully it will!) then add `Closes #<issue number>` here.
-->

**Check #1510 before merging**

### :microscope: Tests

<!--
- Document any manual tests that you have carried out (*e.g.* deploying a new SHM and/or SRE) and confirm which commit you did this with
- Note that automated tests will be run as part of the CI process and will block this PR until they pass.
- Additionally, a successful code review may be required before this PR can be merged.
- If this pull request only changed documentation you can simply state 'Documentation only' here.
-->

JimMadge committed:
Migrate pyproject to PyPA standard and hatch

JimMadge commented:
@jemrobinson This is not ready for review yet but would like your thoughts.

Especially on how to merge this (if we want to) without disrupting other work on the pulumi branch.

JimMadge committed:
Remove flake8 configuration dotfile

JimMadge committed:
Correct lint env dependencies

JimMadge committed:
Set max line length to 120

In line with previous practice.

JimMadge committed:
Use hatch for python package linting

jemrobinson commented on data_safe_haven/__about__.py:
```diff
@@ -0,0 +1 @@
+__version__ = "v0.4.0"
```
At the moment this is in `pyproject.toml`. Is that a standard location? If we also need it in `__about__.py` we can do

```
import pkg_resources
__version__ = pkg_resources.get_distribution("data-safe-haven").version
```

jemrobinson replied:
See eg (https://stackoverflow.com/questions/458550/standard-way-to-embed-version-into-python-package) and (https://packaging.python.org/en/latest/specifications/declaring-project-metadata/#declaring-project-metadata) .

JimMadge replied:
In `pyproject.toml` the version is now dynamic metadata

https://github.com/alan-turing-institute/data-safe-haven/blob/f01446eabc6237930d10e49f61d59d1687887e9d/pyproject.toml#L7

sourced from `__about__.py`

https://github.com/alan-turing-institute/data-safe-haven/blob/f01446eabc6237930d10e49f61d59d1687887e9d/pyproject.toml#L52-L53

So it is only defined in one place (and you can use hatch to update the version).
This is [#1 of the options here](https://packaging.python.org/en/latest/guides/single-sourcing-package-version/), just for hatch rather than setuptools.

I think the things you have linked won't work as they rely on setuptools, which isn't being used here.

JimMadge replied:
We can import (or move `version` to `__init__.py`, maybe renamed `__version__`) to get the value in the code.

jemrobinson replied:
Looks like a relatively common standard is to have `version.py` that contains:

```
__version_info__ = ('1', '2', '3')
__version__ = '.'.join(__version_info__)
```

then `__init__.py` has `from .version import __version__, __version_info__`.

Presumably this will still work with `pyproject.toml` dynamic while providing the PEP-recommended `<module>.__version__` string? As far as I can tell, `__about__.py` is not specified anywhere.

JimMadge replied:
Fixed in 18cba63d

jemrobinson commented on pyproject.toml:
```diff
@@ -1,57 +1,135 @@
+requires-python = ">=3.10"
+license = "BSD-3-Clause"
+dependencies = [
+  "azure-core~=1.26.0",
```
Is `~=` equivalent to the `^=` that poetry uses?

JimMadge replied:
Yes. Even better, it is the [PEP440](https://peps.python.org/pep-0440/#version-specifiers) and [PEP508](https://peps.python.org/pep-0508/) way of specifying versions/requirements.

jemrobinson commented on pyproject.toml:
```diff
@@ -1,57 +1,135 @@
+
+[tool.ruff]
+line-length = 120
+select = [
+  "A",
+  "ARG",
+  "B",
+  "C",
+  "DTZ",
+  "E",
+  "EM",
+  "F",
+  "FBT",
+  "I",
+  "ICN",
+  "ISC",
+  "N",
+  "PLC",
+  "PLE",
+  "PLR",
+  "PLW",
+  "Q",
+  "RUF",
+  "S",
+  "T",
+  "TID",
+  "UP",
+  "W",
+  "YTT",
```
Can you add comments about what these mean?

JimMadge replied:
Done in be27d032.

These are inherited from Hatch's opinionated, sensible defaults.
We may want to remove some, but I've found there are some handy additions like more strictly enforcing conventions, forcing the use of more modern patterns/features.

jemrobinson commented on pyproject.toml:
```diff
@@ -1,57 +1,135 @@
+  "YTT",
+]
+ignore = [
+  # Allow non-abstract empty methods in abstract base classes
+  "B027",
```
Do we need this?

JimMadge replied:
I don't think so, let's remove it.

JimMadge replied:
Done in b22356de

jemrobinson commented on pyproject.toml:
```diff
@@ -1,57 +1,135 @@
+ignore = [
+  # Allow non-abstract empty methods in abstract base classes
+  "B027",
+  # Allow boolean positional values in function calls, like `dict.get(... True)`
+  "FBT003",
```
Do we need this?

JimMadge replied:
There are quite a few instances of this rule being broken.

I think there is a good argument to follow this rule. `do_something(..., True)` is not clear and could mean `do_something(..., verbose=True)` or `(do_something(..., dry_run=True)`.
There is a rule to stop _us_ writing functions with position boolean arguments.

JimMadge replied:
Removed (at least for now) in f8a5e6f9

jemrobinson reviewed (changes requested):
Mostly happy but a few questions on details. Also, we should check for compatibility with #1509 which adds some dependencies.

JimMadge commented:
> Mostly happy but a few questions on details. Also, we should check for compatibility with https://github.com/alan-turing-institute/data-safe-haven/issues/1509 which adds some dependencies.

:+1: Can we agree to hold work on the `python-migration` branch after #1509 merges to let me get linting passing on this branch and merge it?

jemrobinson commented:
I think this and #1509 can be merged in either order. Just need to be aware that they are inter-dependent.

JimMadge commented:
No, because I'll need to fix all the black and ruff errors to not break CI in main. That will involve touching (potentially) every Python file in the repo.

That's why I want other work on this branch to stop before I start fixing the linting errors. Resolving merge conflicts will be too difficult otherwise.

JimMadge committed:
Remove ignore for B027

JimMadge committed:
Add comments for Ruff rules

JimMadge committed:
Improve version specification

JimMadge committed:
Remove ignore for FBT003

JimMadge committed:
Fix version specifier

JimMadge committed:
Run black and ruff

JimMadge committed:
Remove unfixable for F401

JimMadge committed:
Remove unused modules from mypy overrides

JimMadge committed:
Fix ruff errors in types.py

JimMadge committed:
Fix Ruff errors in logger.py

JimMadge committed:
Add missing version.py

jemrobinson commented on data_safe_haven/administration/users/__init__.py:
```diff
@@ -1,3 +1,3 @@
-from .user_handler import UserHandler
+from data_safe_haven.administration.users.user_handler import UserHandler
```
Why is this type of import better? We're in `__init__.py` here and defining what should be exported from this module. If anything *isn't* in this directory that's a problem, and I feel like this syntax makes it harder to verify that.

JimMadge replied:
That is [this rule](https://beta.ruff.rs/docs/rules/relative-imports/). I like it in general. This is a case where it looks odd though.

We can [configure it to allow relative imports of siblings](https://beta.ruff.rs/docs/settings/#flake8-tidy-imports-ban-relative-imports).

jemrobinson replied:
Happy if you prefer it. I'll let you decide.

JimMadge replied:
Let's allow it. I don't see a strong argument to always require absolute imports.

(If we find one, it is easy to change the configuration and let ruff update them all).

JimMadge replied:
Updated in 741290b0

jemrobinson commented on data_safe_haven/administration/users/active_directory_users.py:
```diff
@@ -1,15 +1,17 @@
+from collections.abc import Sequence
+from typing import Any
+
+from data_safe_haven.administration.users.research_user import ResearchUser
```
This seems to be in the wrong order. If we want to separate same-package imports from same-folder imports, then same-folder ones should come afterwards.

JimMadge replied:
I think this happened because of the comment near the other imports and the adding of the absolute path.

If relative imports are allowed the order can be like before

```Python
import pathlib
from collections.abc import Sequence
from typing import Any

from data_safe_haven.config import Config
from data_safe_haven.external import AzureApi
from data_safe_haven.functions import b64encode
from data_safe_haven.pulumi import PulumiSHMStack
from data_safe_haven.utility import FileReader, Logger

from .research_user import ResearchUser
```

JimMadge replied:
Fixed in 741290b0

jemrobinson commented on data_safe_haven/administration/users/active_directory_users.py:
```diff
@@ -70,11 +66,9 @@ def add(self, new_users: Sequence[ResearchUser]) -> None:
             self.logger.parse(line)
 
-    def list(self, sre_name: Optional[str] = None) -> Sequence[ResearchUser]:
+    def list(self, sre_name: str | None = None) -> Sequence[ResearchUser]:
```
I prefer `Optional[type]` to `type | None` in function signatures because it correctly describes the parameter as being `optional`.

JimMadge replied:
[The rule](https://beta.ruff.rs/docs/rules/non-pep604-annotation/) here is enforcing using the new union syntax rather than `typing.Union` or `typing.Optional`. **I don't think it will distinguish when arguments are optional**.

I would go with the rule (apart from where it breaks Typer) because

- It is more concise
- It is more clear that the type is actually a union of T and None
- An `Optional[T]` is not the same as an optional argument, _e.g._

```Python
def f(x: Optional[int]) -> None:
    if x is None:
        do_something()
    else:
        do_something_else()
```
x is `Optional[int]` but it is not an optional argument.

jemrobinson replied:
I mean that I like the use of `Optional[type]` to indicate that you *do have* an optional argument. It's not a hill I would die on though!

JimMadge replied:
Yep. The linter doesn't distinguish between using `typing.Optional` for optional arguments and in other places. It just prefers the newer union operator syntax. So I don't think there is much hope of enforcing that.

I'll keep the rule here apart from exclusions for Typer.

jemrobinson commented on data_safe_haven/administration/users/azure_ad_users.py:
```diff
@@ -63,11 +57,7 @@ def list(self) -> Sequence[ResearchUser]:
-                    if len(user_details["businessPhones"])
-                    else None
-                ),
+                phone_number=(user_details["businessPhones"][0] if len(user_details["businessPhones"]) else None),
```
Note that lots of these changes will be reverted by `black`

jemrobinson reviewed (changes requested):
`Typer` requires `Optional[type]` rather than `type | None`. Otherwise you get the message:

```
RuntimeError: Type not yet supported: str | None
```

JimMadge committed:
Fix Ruff errors in pulumi_stack.py

JimMadge committed:
Fix Ruff errors in ssl_certificate.py

jemrobinson commented on data_safe_haven/pulumi/dynamic/ssl_certificate.py:
```diff
@@ -115,14 +114,17 @@ def create(self, props: dict[str, Any]) -> CreateResult:
             outs["secret_id"] = kvcert.secret_id
         except Exception as exc:
-            msg = f"Failed to create SSL certificate [green]{props['certificate_secret_name']}[/] for [green]{props['domain_name']}[/].\n{exc!s}"
+            msg = (
```
This might be clearer as `" ".join([...])` rather than needing to remember to leave whitespace in the strings.

JimMadge replied:
I disagree. If we use `" ".join` we can _only_ split on spaces.

I think string literal concatenation is idiomatic and efficient (concatenated at compile time rather than run time).

jemrobinson replied:
Fair enough. There are a couple of places where you lost whitespace while (manually?) doing the splitting. Can you check?

jemrobinson commented on data_safe_haven/pulumi/dynamic/ssl_certificate.py:
```diff
@@ -138,12 +140,14 @@ def delete(self, id_: str, props: dict[str, Any]) -> None:
         except Exception as exc:
-            msg = f"Failed to delete SSL certificate [green]{props['certificate_secret_name']}[/] for [green]{props['domain_name']}[/].\n{exc!s}"
+            msg = (
+                f"Failed to delete SSL certificate [green]{props['certificate_secret_name']}[/]"
```
E.g. see here where necessary whitespace is missing

JimMadge replied:
Fixed in 19bad5a47040fd59008289caa49c40cd31bacd8f

jemrobinson commented on data_safe_haven/pulumi/pulumi_stack.py:
```diff
@@ -198,17 +198,15 @@ def login(self) -> None:
-                    if process.stdout:
-                        self.logger.info(process.stdout.readline().strip())
+                )
+                process.check_returncode()
```
Add a comment that this raises a `CalledProcessError` for a non-zero return code.

JimMadge replied:
Fixed in 4058ee0b

JimMadge committed:
Fix Ruff errors in remote_powershell.py

JimMadge committed:
Fix ruff errors in file_share_file.py

jemrobinson commented on data_safe_haven/pulumi/dynamic/remote_powershell.py:
```diff
@@ -50,13 +50,8 @@ def create(self, props: dict[str, Any]) -> CreateResult:
 
-    def delete(self, id_: str, props: dict[str, Any]) -> None:
-        """The Python SDK does not support configuration deletion"""
-        return
```
We should include at least a comment explaining why `delete` is not needed here (probably because you can't delete the fact that a script has run). Also check whether `DshResourceProvider` requires this method to exist or not.

JimMadge replied:
I wasn't sure exactly the best thing to do here.

- It was clear that delete was not valid
- The parent class `DshResourceProvider` is not an `ABC` and the `delete` method is not an `abstractmethod`. So there is no error if it is not defined when creating a child of `DshResourceProvider`. ([see here](https://docs.python.org/3/library/abc.html?highlight=abc))

Having a closer look though, `DshResourceProvider`'s `update` method does call `delete`

https://github.com/alan-turing-institute/data-safe-haven/blob/4058ee0b00c168da98ede70ebaeadf36b457696e/data_safe_haven/pulumi/dynamic/dsh_resource_provider.py#L67-L76

I can't find this method every being called though.

So, maybe we should do one of

1. Redefine `update` in `RemoteScriptProvider` (seems like not a great idea).
2. Make the appropriate methods of `DshResourceProvider` abstract methods and add a comment to say what to do if `delete` is meaningless.

What do you think @jemrobinson?

In any case, I think this should be a new issue/PR.
I can revert the changes and ignore in CI.

jemrobinson replied:
Captured in #1516

JimMadge committed:
Fix Ruff errors in dsh_resource_provider.py

JimMadge committed:
Fix Ruff errors in compiled_dsc.py

JimMadge committed:
Fix Ruff errors in blob_container_acl.py

JimMadge committed:
Fix Ruff errors in azuread_applications.py

jemrobinson commented on data_safe_haven/pulumi/dynamic/dsh_resource_provider.py:
```diff
@@ -44,26 +44,21 @@ def partial_diff(
     def refresh(props: dict[str, Any]) -> dict[str, Any]:
         return dict(**props)
 
-    def check(self, old_props: dict[str, Any], new_props: dict[str, Any]) -> CheckResult:
```
These signatures can't be changed as they're inherited from `pulumi.dynamic.ResourceProvider`.

JimMadge replied:
Fixed in 65280d45

JimMadge committed:
Fix Ruff errors in virtual_machine.py

JimMadge committed:
Fix Ruff errors in sre_software_repositories.py

JimMadge committed:
Fix Ruff errors in sre_networking.py

JimMadge committed:
Fix Ruff errors in sre_remote_desktop.py

JimMadge committed:
Fix Ruff errors in sre_monitoring.py

JimMadge committed:
Fix ARG002 errors

JimMadge committed:
Fix Ruff errors in sre_hedgedoc_server.py

JimMadge committed:
Fix Ruff errors in sre_gitea_server.py

jemrobinson commented on data_safe_haven/pulumi/components/sre_hedgedoc_server.py:
```diff
@@ -258,7 +258,10 @@ def __init__(
                             name="CMD_LDAP_SEARCHFILTER",
-                            value=f"(&(objectClass=user)(memberOf=CN={props.ldap_user_security_group_name},OU=Data Safe Haven Security Groups,{props.ldap_root_dn})(sAMAccountName={{{{username}}}}))",
+                            value=(
+                                f"(&(objectClass=user)(memberOf=CN={props.ldap_user_security_group_name},"
```
This is a really bad string to break up. Brackets, punctuation etc. are all important and need to be easily verified.

Can you break it up into logical pieces?

I'd suggest something like:

```
"(&"
"(objectClass=user)"
f"(memberOf=CN={props.ldap_user_security_group_name},OU=Data Safe Haven Security Groups,{props.ldap_root_dn})"
f"(sAMAccountName={{{{username}}}})"
")"
```

jemrobinson replied:
Or maybe:

```
security_groups_ou = f"Data Safe Haven Security Groups,{props.ldap_root_dn}"
user_security_group_cn = f"{props.ldap_user_security_group_name},OU={security_groups_ou}"
value=f"(&(objectClass=user)(memberOf=CN={user_security_group_cn})(sAMAccountName={{{{username}}}}))"
```

JimMadge replied:
Fixed in e856efa9

JimMadge committed:
Fix Ruff errors in sre_data.py

jemrobinson commented on data_safe_haven/utility/types.py:
```diff
@@ -1,6 +1,6 @@
 
-PathType = Union[str, Path]
+PathType = str | Path
 YamlType = dict[str, dict[str, Any]]
```
I think this type might be wrong. Maybe it should be `dict[str, Any | dict[str, Any]]`? Not sure how helpful that is though.

JimMadge replied:
Fixed in 6af37b02

JimMadge committed:
Fix Ruff errors in sre_application_gateway.py

jemrobinson commented on data_safe_haven/pulumi/components/sre_data.py:
```diff
@@ -355,7 +355,9 @@ def __init__(
                 acl_group="rwx",
                 acl_other="rwx",
-                apply_default_permissions=False,  # due to an Azure bug this also gives ownership of the fileshare to user 65533 (preventing use inside the SRE)
+                # due to an Azure bug this also gives ownership of the
```
It needs to be clear that this comment applies to `apply_default_permissions` not to a) the function in general or b) `acl_other`. I think it's easier to see this as an in-line comment.

JimMadge replied:
:+1: I would say it is clear because it appears in the argument list before the argument. The same way we comment what a line does before the line.

I would rather not have it as an inline comment as it is quite long _and_ on an indented line (I tend to avoid inline comments entirely).
We could change the text to make it more clear what the comment refers to?

JimMadge replied:
Fixed in 362eafde

JimMadge committed:
Fix line length errors

JimMadge committed:
Fix Ruff errors in validators.py

JimMadge committed:
Fix Ruff errors in miscellaneous.py

JimMadge committed:
Fix Ruff errors in azure_postgresql_database.py

JimMadge committed:
Fix Ruff errors in azure_ipv4_range.py

JimMadge committed:
Fix Ruff errors in graph_api.py

JimMadge committed:
Fix Ruff errors in azure_cli.py

JimMadge committed:
Use PEP naming convention for Exceptions

JimMadge committed:
Fix Ruff errors in config.py

JimMadge committed:
Ignore error for print statement

JimMadge committed:
Ignore A003 where it improves clarity

JimMadge committed:
Ignore magic number where it is not resused

JimMadge committed:
Run fmt script

JimMadge commented:
@jemrobinson Some things I found

- I've ignored [A003](https://beta.ruff.rs/docs/rules/builtin-attribute-shadowing/) where breaking the rule improves clarity. I would be happy to remove it altogether. Shadowing builtin names is bad, but when they are methods I think you are unlikely to confused them `set(abc)` vs `myclass.set(abc)`.
- I've ignored an error raised by a `print` call inline. Feels like it is a useful failure as there shouldn't be _too many_ print statements in the code (we may add them for some quick and dirty debugging).
- fea84c13 Ignored but I don't like this code. What is 6 here?
- _I_ would prefer a shorted line length. 80-90 seems good. Some of the long lines are difficult to read.

JimMadge committed:
Fix whitespace in string

jemrobinson commented:
> @jemrobinson Some things I found
> 
>     * I've ignored [A003](https://beta.ruff.rs/docs/rules/builtin-attribute-shadowing/) where breaking the rule improves clarity. I would be happy to remove it altogether. Shadowing builtin names is bad, but when they are methods I think you are unlikely to confused them `set(abc)` vs `myclass.set(abc)`.

Happy to change these method names, but maybe in a separate PR for clarity?
 
>     * I've ignored an error raised by a `print` call inline. Feels like it is a useful failure as there shouldn't be _too many_ print statements in the code (we may add them for some quick and dirty debugging).

I don't think there's a better way to write the version number to the console than this. I guess we could use `sys.stdout.write` but that feels silly.

>     * [fea84c1](https://github.com/alan-turing-institute/data-safe-haven/commit/fea84c13d4a93164ef8116ad7a6636e792f572ad) Ignored but I don't like this code. What is 6 here?

We are expecting there to be six tokens for the following line to work. Happy to replace this with a `try/except` if you'd prefer.

```
ResearchUser(
                        email_address=tokens[4],
                        given_name=tokens[1],
                        phone_number=tokens[3],
                        sam_account_name=tokens[0],
                        surname=tokens[2],
                        user_principal_name=tokens[5],
)
```

> I would prefer a shorted line length. 80-90 seems good. Some of the long lines are difficult to read.

I don't mind longer lines as I use a wide screen and frequently full-screen my code window, but I note that `black` uses 88. If you want to default to this (ie. don't set `line-length` in our `black` config) that's fine with me.

jemrobinson commented on data_safe_haven/administration/users/user_handler.py:
```diff
@@ -81,98 +74,88 @@ def get_usernames(self) -> Dict[str, List[str]]:
-            raise DataSafeHavenUserHandlingException(
-                f"Could not list users.\n{str(exc)}"
-            ) from exc
+            msg = f"Could not list users.\n{exc!s}"
```
`exc!s` is definitely valid, but it is clearer/more common than `str(exc)`? No problem with either, but I want to make sure this is a deliberate choice.

JimMadge replied:
This is actually new to me. So, I kind of dislike it because it looks unusual. However, it is shorter and part of the f-string grammar. I would say keep the rule.

(Also, I wonder if `f"{str(a)}"` actually calls `str()` twice. Once to evaluate the f expression then again for the conversion.)

The rule is [this one](https://beta.ruff.rs/docs/rules/explicit-f-string-type-conversion/).

I think all of these are unnecessary though as the default is to use `str()`. So I think we should remove all of these.

JimMadge replied:
Fixed in ef21e9a6

jemrobinson commented on data_safe_haven/backend/__init__.py:
```diff
@@ -1,4 +1,4 @@
-from .backend import Backend
+from data_safe_haven.backend.backend import Backend
```
Again, I find this import style overly verbose, but if you prefer it that's fine.

JimMadge replied:
Fixed in 741290b0bbc51de7d7222022485b16ee35b5566e

jemrobinson commented on data_safe_haven/backend/backend.py:
```diff
@@ -59,9 +57,8 @@ def create(self) -> None:
-                raise DataSafeHavenAzureException(
-                    f"Storage account '{self.config.backend.storage_account_name}' was not created."
-                )
+                msg = f"Storage account '{self.config.backend.storage_account_name}' was not created."
+                raise DataSafeHavenAzureError(msg)
```
I'm not sure I find this clearer. We create a variable that is only used on one line (the following one) rather than using brackets and a line break. Is it the indentation that's the issue here?

JimMadge replied:
[Keeps tracebacks easy (at least easier) to read](https://beta.ruff.rs/docs/rules/f-string-in-exception/)

That is a new one to me :smile:. Sounds worth keeping.

jemrobinson replied:
This is such a minor issue that I'm not sure I would have bothered putting it into a syntax checker! I don't think it's *wrong* though, so I'm happy to keep it.

jemrobinson commented on data_safe_haven/backend/backend.py:
```diff
@@ -81,32 +78,25 @@ def create(self) -> None:
-            self.config.pulumi.encryption_key_id = pulumi_encryption_key.id.split("/")[
-                -1
-            ]
+            self.config.pulumi.encryption_key_id = pulumi_encryption_key.id.split("/")[-1]
```
I prefer this, but `black` imposed the other syntax. What setting have you changed to make sure it won't do so if I run it again?

JimMadge replied:
Not certain what would be different between our configurations. The black rules are in `pyproject.toml` so everyone _should_ be able to arrive at the same result by running `hatch run lint:fmt`.

jemrobinson commented on data_safe_haven/cli.py:
```diff
@@ -14,19 +14,17 @@
     ] = None,
     verbosity: Annotated[
-        Optional[int],
+        int | None,
```
Typer requires `Optional`

JimMadge replied:
Fixed in 40ae88b9

jemrobinson commented on data_safe_haven/commands/users_unregister_command.py:
```diff
@@ -1,12 +1,11 @@
 from data_safe_haven.administration.users import UserHandler
 from data_safe_haven.config import Config
 from data_safe_haven.exceptions import (
-    DataSafeHavenException,
+    DataSafeHavenError,
 )
```
This should be on one line.

jemrobinson commented on data_safe_haven/config/config.py:
```diff
@@ -43,29 +43,25 @@ def validate(self) -> None:
-                f"Invalid value for 'admin_group_id' ({self.admin_group_id}).\n{str(exc)}"
-            )
+            msg = f"Invalid value for 'admin_group_id' ({self.admin_group_id}).\n{exc!s}"
+            raise ValueError(msg) from exc
```
Should we make these `DataSafeHavenError`s? Maybe `DataSafeHavenInputError` or `DataSafeHavenParameterError`?

JimMadge replied:
I think any of those will work. The correct answer is probably what is most useful for the user.

I would say this is out of scope for this PR though. Can you open an issue?

jemrobinson replied:
Opened #1515

jemrobinson commented on data_safe_haven/exceptions/__init__.py:
```diff
@@ -1,50 +1,50 @@
 
 
-class DataSafeHavenIPRangeException(DataSafeHavenException):
+class DataSafeHavenIPRangeError(DataSafeHavenError):
     pass
 
 
-class DataSafeHavenNotImplementedException(DataSafeHavenInternalException):
+class DataSafeHavenNotImplementedError(DataSafeHavenInternalError):
     pass
 
 
-class DataSafeHavenParameterException(DataSafeHavenException):
+class DataSafeHavenParameterError(DataSafeHavenError):
     pass
 
 
-class DataSafeHavenSSLException(DataSafeHavenException):
+class DataSafeHavenSSLError(DataSafeHavenError):
     pass
 
 
-class DataSafeHavenAzureException(DataSafeHavenCloudException):
+class DataSafeHavenAzureError(DataSafeHavenCloudError):
     pass
 
 
-class DataSafeHavenUserHandlingException(DataSafeHavenInternalException):
+class DataSafeHavenUserHandlingError(DataSafeHavenInternalError):
     pass
 
 
-class DataSafeHavenMicrosoftGraphException(DataSafeHavenAzureException):
+class DataSafeHavenMicrosoftGraphError(DataSafeHavenAzureError):
     pass
 
 
-class DataSafeHavenPulumiException(DataSafeHavenCloudException):
+class DataSafeHavenPulumiError(DataSafeHavenCloudError):
     pass
```
Worth checking whether all these errors are currently used in the code. Some may be unnecessary at the moment.

JimMadge replied:
Definitely worth doing. Do you know if there is a tool to check for unused exceptions?

Again, feels out of scope for the PR?

jemrobinson replied:
I think a quick "grep" will do the job here.

JimMadge replied:
These are all used :+1:.

I have opened an issue about clarifying their use https://github.com/alan-turing-institute/data-safe-haven/issues/1518

jemrobinson commented on data_safe_haven/external/api/azure_api.py:
```diff
@@ -28,38 +29,43 @@
 from azure.mgmt.keyvault import KeyVaultManagementClient
-from azure.mgmt.keyvault.models import AccessPolicyEntry, Permissions
-from azure.mgmt.keyvault.models import Sku as KeyVaultSku
 from azure.mgmt.keyvault.models import (
+    AccessPolicyEntry,
+    Permissions,
     Vault,
     VaultCreateOrUpdateParameters,
     VaultProperties,
 )
+from azure.mgmt.keyvault.models import Sku as KeyVaultSku
```
These could/should be in one include block. What's the reason for separating them?

JimMadge replied:
You tell me 2143eec87 :wink:.

A little disapointed the autoformatters don't concatenate these.

JimMadge replied:
Looks like it is the isort style. I suppose separating aliased imports.

JimMadge replied:
Could change this behaviour with [combine-as-imports](https://beta.ruff.rs/docs/settings/#isort-combine-as-imports). Do you have a strong opinion?

I don't mind too much but would lean towards the default for simplicity's sake.

jemrobinson replied:
I'd like to see these combined if there's a set of arguments that will make this happen.

JimMadge replied:
Fixed in 6a73875b

jemrobinson commented on data_safe_haven/external/api/azure_api.py:
```diff
@@ -28,38 +29,43 @@
 from azure.mgmt.storage import StorageManagementClient
 from azure.mgmt.storage.models import (
     BlobContainer,
-    Kind as StorageAccountKind,
     PublicAccess,
-    Sku as StorageAccountSku,
     StorageAccount,
     StorageAccountCreateParameters,
     StorageAccountKey,
     StorageAccountListKeysResult,
 )
+from azure.mgmt.storage.models import (
+    Kind as StorageAccountKind,
+)
+from azure.mgmt.storage.models import (
+    Sku as StorageAccountSku,
+)
```
These could/should be in one include block. What's the reason for separating them?

JimMadge replied:
Same as above.

jemrobinson commented on data_safe_haven/functions/strings.py:
```diff
@@ -47,20 +48,15 @@ def random_letters(length: int) -> str:
 
 
-def truncate_tokens(tokens: Sequence[str], max_length: int) -> List[str]:
+def truncate_tokens(tokens: Sequence[str], max_length: int) -> list[str]:
```
```suggestion
def truncate_tokens(tokens: Sequence[str], max_length: int) -> list[str]:
    """
    Recursively remove the final character from the longest strings in the input.
    Terminate when the total length of all strings is no greater than max_length.
    For example:
        truncate_tokens(["the", "quick", "fox"], 6) -> ["th", "qu", "fo"]
    """
```

jemrobinson commented on data_safe_haven/pulumi/common/transformations.py:
```diff
@@ -1,17 +1,16 @@
 """Common transformations needed when manipulating Pulumi resources"""
 # Standard library imports
```
Delete this line if there are no stdlib imports

jemrobinson commented on data_safe_haven/pulumi/components/shm_firewall.py:
```diff
@@ -1,6 +1,5 @@
 """Pulumi component for SHM traffic routing"""
 # Standard library import
```
Delete this line.

jemrobinson commented on data_safe_haven/pulumi/components/shm_firewall.py:
```diff
@@ -49,15 +44,16 @@ def __init__(
 
         # Important IP addresses
-        external_dns_resolver = "168.63.129.16"  # https://docs.microsoft.com/en-us/azure/virtual-network/what-is-ip-address-168-63-129-16
+        external_dns_resolver = (
```
Move comment out of inline?

JimMadge replied:
Fixed in f6ff62e5.

jemrobinson commented on data_safe_haven/pulumi/components/sre_monitoring.py:
```diff
@@ -1,6 +1,5 @@
 """Pulumi component for SHM monitoring"""
 # Standard library import
```
Delete this line.

jemrobinson commented on data_safe_haven/utility/logger.py:
```diff
@@ -67,7 +65,7 @@ class LogLevelHighlighter(RegexHighlighter):
 
     base_style = "logging.level."
-    highlights = [
+    highlights: ClassVar[list[str]] = [
```
This should be a tuple instead of a list. That should avoid the need for `ClassVar` here.

jemrobinson reviewed (changes requested):
- Import ordering is broken: imports from local folder should come after first party imports (see eg. https://pycqa.github.io/isort/docs/configuration/custom_sections_and_ordering.html)
- In several places you've changed syntax/linebreaks that were previously imposed by black. Have you run `black` recently? Does it make any changes?
- Typer needs `Optional[type]` instead of `type | None`. I also think this form looks better in function signatures where you're indicated that an argument is `optional`.

JimMadge committed:
Improve subprocess output handling

JimMadge committed:
Remove unnecessary f-string conversions

JimMadge committed:
Remove __about__.py

JimMadge committed:
Allow relative imports of siblings

JimMadge committed:
Remove remaining standard library comment

JimMadge committed:
Add docstring for truncate_tokens

Co-authored-by: James Robinson <james.em.robinson@gmail.com>

JimMadge committed:
Enable black string normalisation

JimMadge committed:
Update linting workflow to run hatch style script

JimMadge committed:
Move inline comment

Co-authored-by: James Robinson <james.em.robinson@gmail.com>

JimMadge committed:
Use typing.Optional for Typer annotations

Typer does not support newer union syntax
https://github.com/tiangolo/typer/pull/522

JimMadge committed:
Explain that check_returncode may raise exception

Co-authored-by: James Robinson <james.em.robinson@gmail.com>

jemrobinson commented on data_safe_haven/administration/users/user_handler.py:
```diff
@@ -62,7 +62,7 @@ def add(self, users_csv_path: pathlib.Path) -> None:
             # Commit changes
             self.active_directory_users.add(users)
         except Exception as exc:
-            msg = f"Could not add users from '{users_csv_path}'.\n{exc!s}"
```
Won't you get `exception.__repr__` instead of `exception.__str__` by doing this?

JimMadge replied:
Nope, the default conversion is `str()`.

JimMadge committed:
Tidy LDAP search filter

Co-authored-by: James Robinson <james.em.robinson@gmail.com>

JimMadge committed:
Clarify comment

Co-authored-by: James Robinson <james.em.robinson@gmail.com>

JimMadge committed:
Use combine-as-imports isort option

jemrobinson commented on data_safe_haven/pulumi/dynamic/azuread_application.py:
```diff
@@ -61,31 +57,28 @@ def create(self, props: Dict[str, Any]) -> CreateResult:
         )
 
-    def delete(self, id_: str, props: Dict[str, Any]) -> None:
+    def delete(self, props: dict[str, Any]) -> None:
```
```suggestion
    def delete(self, id_: str, props: dict[str, Any]) -> None:
```

jemrobinson commented on data_safe_haven/pulumi/dynamic/azuread_application.py:
```diff
@@ -61,31 +57,28 @@ def create(self, props: Dict[str, Any]) -> CreateResult:
-        id_: str,
-        old_props: Dict[str, Any],
-        new_props: Dict[str, Any],
+        old_props: dict[str, Any],
```
```suggestion
        id_: str,
        old_props: dict[str, Any],
```

jemrobinson commented on data_safe_haven/pulumi/dynamic/blob_container_acl.py:
```diff
@@ -63,15 +62,14 @@ def create(self, props: Dict[str, Any]) -> CreateResult:
         )
 
-    def delete(self, id_: str, props: Dict[str, Any]) -> None:
+    def delete(self, props: dict[str, Any]) -> None:
```
```suggestion
    def delete(self, id_: str, props: dict[str, Any]) -> None:
```

jemrobinson commented on data_safe_haven/pulumi/dynamic/blob_container_acl.py:
```diff
@@ -82,16 +80,13 @@ def delete(self, id_: str, props: Dict[str, Any]) -> None:
-        id_: str,
-        old_props: Dict[str, Any],
-        new_props: Dict[str, Any],
+        old_props: dict[str, Any],
```
```suggestion
        id_: str,
        old_props: dict[str, Any],
```

jemrobinson commented on data_safe_haven/pulumi/dynamic/compiled_dsc.py:
```diff
@@ -53,15 +52,14 @@ def create(self, props: Dict[str, Any]) -> CreateResult:
         )
 
-    def delete(self, id_: str, props: Dict[str, Any]) -> None:
+    def delete(self) -> None:
```
```suggestion
    def delete(self, id_: str, props: dict[str, Any]) -> None:
```

jemrobinson commented on data_safe_haven/pulumi/dynamic/compiled_dsc.py:
```diff
@@ -53,15 +52,14 @@ def create(self, props: Dict[str, Any]) -> CreateResult:
-        id_: str,
-        old_props: Dict[str, Any],
-        new_props: Dict[str, Any],
+        old_props: dict[str, Any],
```
```suggestion
        id_: str,
        old_props: dict[str, Any],
```

jemrobinson commented on data_safe_haven/pulumi/dynamic/file_share_file.py:
```diff
@@ -96,17 +93,16 @@ def create(self, props: Dict[str, Any]) -> CreateResult:
         )
 
-    def delete(self, id_: str, props: Dict[str, Any]) -> None:
+    def delete(self, props: dict[str, Any]) -> None:
```
```suggestion
    def delete(self, id_: str, props: dict[str, Any]) -> None:
```

jemrobinson commented on data_safe_haven/pulumi/dynamic/file_share_file.py:
```diff
@@ -118,15 +114,13 @@ def delete(self, id_: str, props: Dict[str, Any]) -> None:
-        id_: str,
-        old_props: Dict[str, Any],
-        new_props: Dict[str, Any],
+        old_props: dict[str, Any],
```
```suggestion
        id_: str,
        old_props: dict[str, Any],
```

jemrobinson commented on data_safe_haven/pulumi/dynamic/remote_powershell.py:
```diff
@@ -50,15 +48,10 @@ def create(self, props: Dict[str, Any]) -> CreateResult:
-        id_: str,
-        old_props: Dict[str, Any],
-        new_props: Dict[str, Any],
+        old_props: dict[str, Any],
```
```suggestion
        id_: str,
        old_props: dict[str, Any],
```

jemrobinson commented on data_safe_haven/pulumi/dynamic/ssl_certificate.py:
```diff
@@ -125,15 +112,17 @@ def create(self, props: Dict[str, Any]) -> CreateResult:
         )
 
-    def delete(self, id_: str, props: Dict[str, Any]) -> None:
+    def delete(self, props: dict[str, Any]) -> None:
```
```suggestion
    def delete(self, id_: str, props: dict[str, Any]) -> None:
```

jemrobinson commented on data_safe_haven/pulumi/dynamic/ssl_certificate.py:
```diff
@@ -149,15 +138,16 @@ def delete(self, id_: str, props: Dict[str, Any]) -> None:
-        id_: str,
-        old_props: Dict[str, Any],
-        new_props: Dict[str, Any],
+        old_props: dict[str, Any],
```
```suggestion
        id_: str,
        old_props: dict[str, Any],
```

JimMadge committed:
Restore required arguments for Pulumi 'callbacks'

JimMadge committed:
Add missing id_ argument

JimMadge committed:
Remove unused YamlType

jemrobinson commented on data_safe_haven/cli.py:
```diff
@@ -1,32 +1,27 @@
-        typer.Option(
-            "--output", "-o", resolve_path=True, help="Path to an output log file"
-        ),
+        pathlib.Path | None,
```
```suggestion
        Optional[pathlib.Path],
```

jemrobinson commented on data_safe_haven/cli.py:
```diff
@@ -1,32 +1,27 @@
     ] = None,
     verbosity: Annotated[
-        Optional[int],
+        int | None,
```
```suggestion
        Optional[int],
```

jemrobinson commented on data_safe_haven/cli.py:
```diff
@@ -36,16 +31,14 @@ def callback(
-        typer.Option(
-            "--version", "-V", help="Display the version of this application."
-        ),
+        bool | None,
```
```suggestion
        Optional[bool],
```

JimMadge committed:
Run linting on python-migration branch

jemrobinson commented on data_safe_haven/commands/typer_init.py:
```diff
@@ -1,18 +1,16 @@
-# Local imports
 from data_safe_haven.functions import validate_aad_guid
+
 from .initialise_command import InitialiseCommand
```
Do we want this to be `from data_safe_haven.commands.initialise_command import InitialiseCommand`?

JimMadge replied:
I think they are the same, this file is in the `data_safe_haven.commands` module. And we are allowing relative imports of siblings now.

jemrobinson commented on data_safe_haven/commands/typer_init.py:
```diff
@@ -1,18 +1,16 @@
 def initialise_command(
     admin_group: Annotated[
-        Optional[str],
+        str | None,
```
```suggestion
        Optional[str], # noqa: UP007
```

jemrobinson commented on data_safe_haven/commands/typer_init.py:
```diff
@@ -21,23 +19,23 @@ def initialise_command(
     ] = None,
     location: Annotated[
-        Optional[str],
+        str | None,
```
```suggestion
        Optional[str], # noqa: UP007
```

jemrobinson commented on data_safe_haven/commands/typer_init.py:
```diff
@@ -21,23 +19,23 @@ def initialise_command(
     ] = None,
     name: Annotated[
-        Optional[str],
+        str | None,
```
```suggestion
        Optional[str], # noqa: UP007
```

jemrobinson commented on data_safe_haven/commands/typer_init.py:
```diff
@@ -21,23 +19,23 @@ def initialise_command(
     ] = None,
     subscription: Annotated[
-        Optional[str],
+        str | None,
```
```suggestion
        Optional[str],  # noqa: UP007
```

jemrobinson commented on data_safe_haven/config/__init__.py:
```diff
@@ -1,5 +1,5 @@
-from .config import Config
 from .backend_settings import BackendSettings
+from .config import Config
```
```suggestion
from data_safe_haven.config.backend_settings import BackendSettings
from data_safe_haven.config.config import Config
```

?

JimMadge replied:
I think there are the same, this file is in the `data_safe_haven.commands` module. And we are allowing relative imports of siblings now.

jemrobinson commented on data_safe_haven/pulumi/components/sre_networking.py:
```diff
@@ -259,7 +222,10 @@ def __init__(
                 network.SecurityRuleArgs(
                     access=network.SecurityRuleAccess.ALLOW,
-                    description="Allow LDAP client requests over TCP (see https://devopstales.github.io/linux/pfsense-ad-join/ for details).",
+                    description=(
+                        "Allow LDAP client requests over TCP"
+                        " (see https://devopstales.github.io/linux/pfsense-ad-join/ for details)."
+                    ),
```
```suggestion
                    description=(
                        "Allow LDAP client requests over TCP. "
                        "See https://devopstales.github.io/linux/pfsense-ad-join/ for details."
                    ),
```

jemrobinson commented on data_safe_haven/pulumi/components/sre_research_desktop.py:
```diff
@@ -1,22 +1,20 @@
 )
-from .virtual_machine import LinuxVMProps, VMComponent
+
+from .components.virtual_machine import LinuxVMProps, VMComponent
```
```suggestion
from data_safe_haven.pulumi.components.virtual_machine import LinuxVMProps, VMComponent
```
?

JimMadge replied:
Fixed in 564e536f

JimMadge committed:
Fix Typer annotations

Typer does not support newer union syntax
https://github.com/tiangolo/typer/pull/522

Co-authored-by: James Robinson <james.em.robinson@gmail.com>

JimMadge committed:
Improve security rule description

Co-authored-by: James Robinson <james.em.robinson@gmail.com>

JimMadge committed:
Correct import path

JimMadge committed:
Use next() to lazily fetch first list element

JimMadge committed:
Use default black line length, ignore in ruff

jemrobinson committed:
:rotating_light: Increase spacing before comment

jemrobinson committed:
:bug: Fix use of subprocess ENV

jemrobinson committed:
:bug: Fix as_dict function which was broken by variable name change

jemrobinson committed:
:bug: Fix missing timezone import

jemrobinson committed:
:bug: Remove incorrect SHM name string parameter from SHM components

jemrobinson committed:
:bug: SHMUpdateServersComponent should take stack_name not shm_name as an argument

jemrobinson committed:
:truck: Better variable name

jemrobinson committed:
:bug: Separate SHM and SRE stack YAML files

jemrobinson committed:
:rotating_light: Remove empty line

jemrobinson committed:
:bug: Remove incorrect SHM/SRE name string parameter from SRE components

jemrobinson committed:
:rotating_light: Fix README

jemrobinson reviewed (approved):
Tested by redeploying an SHM and SRE.

jemrobinson committed:
:bug: Removed incorrect parameters from SREGiteaServerComponent and SREHedgeDocServerComponent calls

jemrobinson committed:
:recycle: Move sre_name from component constructor to props for SREResearchDesktopComponent

jemrobinson committed:
:recycle: Move sre_name from component constructor to props for SRENetworkingComponent

jemrobinson committed:
:see_no_evil: Add ruff cache to gitignore
